components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/validators/JDBCScopeValidator.java

```diff
@@ -224,22 +239,117 @@ private boolean validateScope(String[] requestedScopes, AuthenticatedUser user)
+        if (serviceProvider != null) {
+            ClaimConfig claimConfig = serviceProvider.getClaimConfig();
+            if (claimConfig != null) {
+                return claimConfig.isAlwaysSendMappedLocalSubjectId();
+            } else {
+                throw new IdentityOAuth2Exception("Unable to find claim configuration for service provider of client " +
+                        "id " + clientId);
+            }
+        } else {
+            throw new IdentityOAuth2Exception("Unable to find service provider for client id " + clientId);
+        }
+    }
+
+    private String[] getUserRolesForNotAssociatedFederatedUser(AuthenticatedUser user)
+            throws IdentityOAuth2Exception {
+
+        List<String> userRolesList = new ArrayList<>();
+        IdentityProvider identityProvider =
+                OAuth2Util.getIdentityProvider(user.getFederatedIdPName(), user.getTenantDomain());
+        /*
+        Values of Groups consists unmapped federated roles, mapped local roles and Internal/everyone corresponding to
+        authenticated user.
+        Role mapping consists mapped federated roles with local roles corresponding to IDP.
+        By cross checking federated role mapped local roles and values of groups we can filter valid local roles which
+        mapped to the federated role of authenticated user.
+         */
+        List<String> valuesOfGroups = getValuesOfGroupsFromUserAttributes(user.getUserAttributes());
+        if (CollectionUtils.isNotEmpty(valuesOfGroups)) {
+            for (RoleMapping roleMapping : identityProvider.getPermissionAndRoleConfig().getRoleMappings()) {
+                if (roleMapping != null && roleMapping.getLocalRole() != null) {
+                    if (valuesOfGroups.contains(roleMapping.getLocalRole().getLocalRoleName())) {
+                        userRolesList.add(roleMapping.getLocalRole().getLocalRoleName());
+                    }
+                }
+            }
+        }
+        // By default we provide Internal/everyone role for all users.
+        String internalEveryoneRole = OAuth2Util.getInternalEveryoneRole(user);
+        if (StringUtils.isNotBlank(internalEveryoneRole)) {
+            userRolesList.add(internalEveryoneRole);
+        }
+        return userRolesList.toArray(new String[0]);
+    }
+
+    /**
+     * Get groups params Roles from User attributes.
+     *
+     * @param userAttributes User Attributes
+     * @return User attribute map
+     */
+    private List<String> getValuesOfGroupsFromUserAttributes(Map<ClaimMapping, String> userAttributes) {
+
+        if (MapUtils.isNotEmpty(userAttributes)) {
+            for (Map.Entry<ClaimMapping, String> entry : userAttributes.entrySet()) {
+                if (StringUtils.equals(entry.getKey().getRemoteClaim().getClaimUri(), OAuth2Constants.GROUPS)) {
+                    return Arrays.asList(entry.getValue().split(Pattern.quote(ATTRIBUTE_SEPARATOR)));
+                }
+            }
```

Comment:
Shall we extract this out as a separate method since the same logic is used at L311?

Reply:
It's a separate private method, and L311 serves different logic. Please correct me if I am wrong.